mcnode/api.go

```diff
@@ -328,6 +327,81 @@ func (node *Node) httpDelete(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprintln(w, count)
 }
 
+// datastore interface
+type DataObject struct {
+	Data []byte `json:"data"`
+}
+
+// Get /data/get/{objectId}
+// Retrieves a data object from the datastore
+func (node *Node) httpGetData(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	key58 := vars["objectId"]
+	key, err := multihash.FromB58String(key58)
+	if err != nil {
+		apiError(w, http.StatusBadRequest, err)
+		return
+	}
+
+	data, err := node.ds.Get(Key(key))
+	if err != nil {
+		apiError(w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if data == nil {
+		apiError(w, http.StatusNotFound, UnknownObject)
+		return
+	}
+
+	dao := DataObject{data}
+	err = json.NewEncoder(w).Encode(dao)
+	if err != nil {
+		log.Printf("Error writing response body: %s", err.Error())
+	}
+}
+
+// POST /data/put
+// DATA: A stream of json-encoded data objects
+// Puts a batch of objects to the datastore
+// returns a stream of object ids (B58 encoded content multihashes)
+func (node *Node) httpPutData(w http.ResponseWriter, r *http.Request) {
+	rbody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("http/data/put: Error reading request body: %s", err.Error())
+		return
+	}
+
+	dec := json.NewDecoder(bytes.NewReader(rbody))
+	batch := make([][]byte, 0)
+
+	var dao DataObject
+loop:
+	for {
+		err = dec.Decode(&dao)
+		switch {
+		case err == io.EOF:
+			break loop
+		case err != nil:
+			apiError(w, http.StatusBadRequest, err)
+			return
+		default:
+			batch = append(batch, dao.Data)
```

Comment:
so we're building up the batch in memory then flushing to datastore on EOF; 

- can this OOM? what's the batch size limit?
- what are the error semantics when some of the batch fails to insert? looks like any errors from datastore return a 500 but some of the objects may have been committed already -- this may be OK given that inserts are basically idempotent, but I would prefer more granularity

Basically I'm not sure that we really need to batch in memory given that the LSM DB design explicitly does this inside their _memtable_ structure -- we should let them do this for us. This way we can also pipeline decoding and inserts and return a more meaningful success/error stream. Seems like adding backpressure to a channel is pretty straightforward.